### ISSUE TITLE
0.9.2: internalnotification uses kdb_long_double_t without checking

### ISSUE THREAD

**Issue by haraldg** [ElektraInitiative/libelektra#3450]
Hi,

this problem has already been addressed once in 8743cdc06585f4d867b936337f92568a830ddbe1 but apparently got reintroduced.

## Steps to Reproduce the Problem
Try compiling elektra for an armel based platform.

## Expected Result
Elektra compiles successfully.

## Actual Result
Compilation fails with this error message:
```
In file included from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:54,
                 from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:458:
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:454:14: error: unknown type name 'kdb_long_double_t'; did you mean 'kdb_double_t'?
 #define TYPE kdb_long_double_t
              ^~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/include/macros/type_create_to_value.h:101:58: note: in definition of macro 'TYPE_CONVERSION_SIGNATURE'
  int NAME_MACRO (TYPE_NAME) (const Key * KEY_PARAM_NAME, TYPE * VARIABLE_PARAM_NAME)
                                                          ^~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/include/macros/type_create_to_value.h:122:28: note: in expansion of macro 'TYPE'
 TYPE_CONVERSION_SIGNATURE (TYPE, TYPE_NAME, NAME_MACRO)
                            ^~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h: In function 'elektraInternalnotificationConvertCallbackKdbLongDouble':
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:454:14: error: unknown type name 'kdb_long_double_t'; did you mean 'kdb_double_t'?
 #define TYPE kdb_long_double_t
              ^~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:59:2: note: in expansion of macro 'TYPE'
  TYPE * variable = (TYPE *) ctx->variable;
  ^~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:454:14: error: 'kdb_long_double_t' undeclared (first use in this function); did you mean 'kdb_double_t'?
 #define TYPE kdb_long_double_t
              ^~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:59:21: note: in expansion of macro 'TYPE'
  TYPE * variable = (TYPE *) ctx->variable;
                     ^~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:454:14: note: each undeclared identifier is reported only once for each function it appears in
 #define TYPE kdb_long_double_t
              ^~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:59:21: note: in expansion of macro 'TYPE'
  TYPE * variable = (TYPE *) ctx->variable;
                     ^~~~
In file included from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:458:
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:59:27: error: expected expression before ')' token
  TYPE * variable = (TYPE *) ctx->variable;
                           ^
In file included from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/include/kdbplugin.h:15,
                 from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/include/kdbnotificationinternal.h:18,
                 from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.h:14,
                 from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:10:
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:43:82: warning: implicit declaration of function 'elektraInternalnotificationConvertKdbLongDouble'; did you mean 'elektraInternalnotificationConvertKdbDouble'? [-Wimplicit-function-declaration]
 #define INTERNALNOTIFICATION_CONVERSION_FUNCTION_NAME(TYPE_NAME) ELEKTRA_CONCAT (elektraInternalnotificationConvert, TYPE_NAME)
                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/include/kdbmacros.h:21:31: note: in definition of macro 'ELEKTRA_CONCAT2'
 #define ELEKTRA_CONCAT2(X, Y) X##Y
                               ^
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:43:66: note: in expansion of macro 'ELEKTRA_CONCAT'
 #define INTERNALNOTIFICATION_CONVERSION_FUNCTION_NAME(TYPE_NAME) ELEKTRA_CONCAT (elektraInternalnotificationConvert, TYPE_NAME)
                                                                  ^~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:60:7: note: in expansion of macro 'INTERNALNOTIFICATION_CONVERSION_FUNCTION_NAME'
  if (!INTERNALNOTIFICATION_CONVERSION_FUNCTION_NAME (TYPE_NAME) (key, variable))
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:458:
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h: At top level:
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:454:14: error: unknown type name 'kdb_long_double_t'; did you mean 'kdb_double_t'?
 #define TYPE kdb_long_double_t
              ^~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:47:82: note: in definition of macro 'INTERNALNOTIFICATION_REGISTER_SIGNATURE'
  int INTERNALNOTIFICATION_REGISTER_NAME (TYPE_NAME) (Plugin * handle, Key * key, TYPE * variable)
                                                                                  ^~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/macros/add_type.h:69:42: note: in expansion of macro 'TYPE'
 INTERNALNOTIFICATION_REGISTER_SIGNATURE (TYPE, TYPE_NAME)
                                          ^~~~
In file included from /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:10:
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c: In function 'elektraInternalnotificationGet':
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.h:28:55: error: 'elektraInternalnotificationRegisterKdbLongDouble' undeclared (first use in this function); did you mean 'elektraInternalnotificationRegisterKdbDouble'?
 #define INTERNALNOTIFICATION_REGISTER_NAME(TYPE_NAME) elektraInternalnotificationRegister##TYPE_NAME
                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.h:28:55: note: in definition of macro 'INTERNALNOTIFICATION_REGISTER_NAME'
 #define INTERNALNOTIFICATION_REGISTER_NAME(TYPE_NAME) elektraInternalnotificationRegister##TYPE_NAME
                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/src/plugins/internalnotification/internalnotification.c:540:54: note: in expansion of macro 'INTERNALNOTIFICATION_EXPORT_FUNCTION'
    INTERNALNOTIFICATION_EXPORT_FUNCTION (KdbDouble), INTERNALNOTIFICATION_EXPORT_FUNCTION (KdbLongDouble),
                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[5]: *** [src/plugins/internalnotification/CMakeFiles/elektra-internalnotification-objects.dir/build.make:80: src/plugins/internalnotification/CMakeFiles/elektra-internalnotification-objects.dir/internalnotification.c.o] Error 1
make[5]: Leaving directory '/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/build'
make[4]: *** [CMakeFiles/Makefile2:5197: src/plugins/internalnotification/CMakeFiles/elektra-internalnotification-objects.dir/all] Error 2
make[4]: Leaving directory '/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/build'
make[3]: *** [Makefile:180: all] Error 2
make[3]: Leaving directory '/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/build'
make[2]: *** [Makefile:450: /home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/build_dir/target-arm_arm926ej-s_musl_eabi/elektra-0.9.2/.built] Error 2
make[2]: Leaving directory '/home/harald/packages/libs/elektra'
time: package/feeds/custom/elektra/compile#41.79#9.32#51.70
make[1]: *** [package/Makefile:113: package/feeds/custom/elektra/compile] Error 2
make[1]: Leaving directory '/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64'
make: *** [/home/harald/openwrt-sdk-mxs_gcc-8.4.0_musl_eabi.Linux-x86_64/include/toplevel.mk:217: package/elektra/compile] Fehler 2
```

## System Information

- Elektra Version: 0.9.2
- OpenWRT SDK for mxs

**Comment by markus2330** [comment]
Thank you for reporting the problem!

I think it was introduced in 65ce08ea23a2269ab4107bcd5de4abfa123f2ea2, a simple ifdef around the code in this commit should fix the problem.

**Comment by mpranj** [comment]
Sorry for the delays. I will postpone this to the next release.

4636bf1acd3763b03aff778638cc4649e7e0d95a might fix it, but I was unable to test the fix due to a broken SDK at the time.

**Comment by mpranj** [comment]
@haraldg I used the following changes to OpenWRT/Elektra's Makefile to test my changes:
https://github.com/mpranj/packages/commit/29b1c629c3722ca3cfe7538bc4e75a86a033c0c7

If you update the Elektra package for OpenWRT you can take a look at my changes. A few plugins have been removed so the Makefile definitely needs to be adapted. Note that my changes were quick&dirty just to verify my changes. They are not polished by any means and I do not intend to open a PR myself, as I have no experience with OpenWRT.

**Comment by haraldg** [comment]
Thanks for the heads up.

I think your test doesn't actually test the changes: You didn't enable internalnotification so you probably never reproduced the actual issue?

Anyway: I put enabling internalnotification on my TODO list and might even bump the elektra version, if I get around to it.

**Comment by mpranj** [comment]
> I think your test doesn't actually test the changes

You're absolutely right, thanks for taking a look. I updated it now, it seems to be working nevertheless.

> might even bump the elektra version

Sounds great! We're releasing a new version this weekend (including this fix). It might make sense that you simply bump it to 0.9.5 next week, when it is available.
